python/ray/autoscaler/_private/fake_multi_node/node_provider.py

```diff
@@ -18,6 +28,177 @@
 FAKE_HEAD_NODE_TYPE = "ray.head.default"
 
+FAKE_DOCKER_DEFAULT_GCS_PORT = 16379
+FAKE_DOCKER_DEFAULT_OBJECT_MANAGER_PORT = 18076
+FAKE_DOCKER_DEFAULT_CLIENT_PORT = 10002
+
+DOCKER_COMPOSE_SKELETON = {
+    "version": "3.9",
+    "services": {},
+    "networks": {
+        "ray_local": {}
+    }
+}
+
+DOCKER_NODE_SKELETON = {
+    "networks": ["ray_local"],
+    "mem_limit": "3000m",
+    "mem_reservation": "3000m",
+    "shm_size": "1200m",
+    "volumes": []
+}
+
+DOCKER_HEAD_CMD = (
+    "bash -c \""
+    "sudo mkdir -p {volume_dir} && "
+    "sudo chmod 777 {volume_dir} && "
+    "touch {volume_dir}/.in_docker && "
+    "sudo chmod 700 ~/.ssh && "
+    "sudo chmod 600 ~/.ssh/authorized_keys && "
+    "sudo chmod 600 ~/ray_bootstrap_key.pem && "
+    "sudo chown ray:users ~/.ssh ~/.ssh/authorized_keys && "
+    "((sudo apt update && sudo apt install -y openssh-server && "
+    "sudo service ssh start) || true) && "
+    "sleep 1 && "
+    "{CMD_ENV}RAY_FAKE_CLUSTER=1 ray start --head "
+    "--autoscaling-config=~/ray_bootstrap_config.yaml "
+    "--object-manager-port=8076 "
+    "--num-cpus {num_cpus} "
+    "--num-gpus {num_gpus} "
+    # "--resources='{resources}' "
+    "--block\"")
+
+DOCKER_WORKER_CMD = (
+    "bash -c \""
+    "sudo mkdir -p {volume_dir} && "
+    "sudo chmod 777 {volume_dir} && "
+    "touch {volume_dir}/.in_docker && "
+    "sudo chmod 700 ~/.ssh && "
+    "sudo chmod 600 ~/.ssh/authorized_keys && "
+    "sudo chown ray:users ~/.ssh ~/.ssh/authorized_keys && "
+    "((sudo apt update && sudo apt install -y openssh-server && "
+    "sudo service ssh start) || true) && "
+    "sleep 1 && "
+    f"ray start --address={FAKE_HEAD_NODE_ID}:6379 "
+    "--object-manager-port=8076 "
+    "--num-cpus {num_cpus} "
+    "--num-gpus {num_gpus} "
+    # "--resources='{resources}' "
+    "--block\"")
+
+
+def host_dir(container_dir: str):
+    """Replace local dir with potentially different host dir.
+
+    E.g. in docker-in-docker environments, the host dir might be
+    different to the mounted directory in the container.
+
+    This method will do a simple global replace to adjust the paths.
+    """
+    ray_tempdir = os.environ.get("RAY_TEMPDIR", None)
+    ray_hostdir = os.environ.get("RAY_HOSTDIR", None)
+
+    if not ray_tempdir or not ray_hostdir:
+        return container_dir
+
+    return container_dir.replace(ray_tempdir, ray_hostdir)
+
+
+def create_node_spec(head: bool,
+                     docker_image: str,
+                     mounted_cluster_dir: str,
+                     mounted_node_dir: str,
+                     num_cpus: int = 2,
+                     num_gpus: int = 0,
+                     resources: Optional[Dict] = None,
+                     env_vars: Optional[Dict] = None,
+                     host_gcs_port: int = 16379,
+                     host_object_manager_port: int = 18076,
+                     host_client_port: int = 10002,
```

Comment:
we can use the global variables here right?

Reply:
Specifically for these three we probably want to keep them configurable via the provider config (like right now) so we can potentially run multiple clusters at the same time.

Reply:
Sorry I meant to use global variables as the default value? Because they're defined above in the file.